src/script/auth/page/OAuthPermissions.tsx

```diff
@@ -0,0 +1,299 @@
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see http://www.gnu.org/licenses/.
+ *
+ */
+
+import React, {useState} from 'react';
+
+import {OAuthBody} from '@wireapp/api-client/lib/oauth/OAuthBody';
+import {OAuthClient} from '@wireapp/api-client/lib/oauth/OAuthClient';
+import {FormattedMessage, useIntl} from 'react-intl';
+import {connect} from 'react-redux';
+import {AnyAction, Dispatch} from 'redux';
+import {container} from 'tsyringe';
+
+import {
+  Button,
+  ButtonVariant,
+  ContainerXS,
+  Text,
+  Paragraph,
+  Box,
+  Link,
+  LinkVariant,
+  COLOR_V2,
+  H2,
+} from '@wireapp/react-ui-kit';
+
+import {Icon} from 'Components/Icon';
+import {AssetRemoteData} from 'src/script/assets/AssetRemoteData';
+import {AssetRepository} from 'src/script/assets/AssetRepository';
+import {KEY} from 'Util/KeyboardUtil';
+import {loadDataUrl} from 'Util/util';
+
+import {Page} from './Page';
+
+import {Config} from '../../Config';
+import {oauthStrings} from '../../strings';
+import {actionRoot} from '../module/action';
+import {bindActionCreators, RootState} from '../module/reducer';
+import * as SelfSelector from '../module/selector/SelfSelector';
+
+interface Props extends React.HTMLProps<HTMLDivElement> {
+  assetRepository?: AssetRepository;
+}
+
+export enum Scope {
+  WRITE_CONVERSATIONS = 'write:conversations',
+  WRITE_CONVERSATIONS_CODE = 'write:conversations_code',
+  READ_SELF = 'read:self',
+  READ_FEATURE_CONFIGS = 'read:feature_configs',
+}
+
+const OAuthPermissionsComponent = ({
+  doLogout,
+  getOAuthApp,
+  selfUser,
+  selfTeamId,
+  assetRepository = container.resolve(AssetRepository),
+  postOauthCode,
+  getSelf,
+  getTeam,
+}: Props & ConnectedProps & DispatchProps) => {
+  const {formatMessage: _} = useIntl();
+  const [teamImage, setTeamImage] = React.useState<string | ArrayBuffer | undefined>(undefined);
+
+  const oauthParams = decodeURIComponent(window.location.search.slice(1))
+    .split('&')
+    .reduce((acc, param) => {
+      const [key, value] = param.split('=');
+      if (key === 'scope') {
+        return {...acc, [key]: value.replaceAll('+', ' ')};
+      }
+      return {...acc, [key]: value};
+    }, {} as OAuthBody);
+  const [oAuthApp, setOAuthApp] = useState<OAuthClient | null>(null);
+  const oAuthScope = oauthParams.scope
+    .split(/\+|%20|\s/)
+    .filter(scope => Object.values(Scope).includes(scope as Scope)) as Scope[];
```

Comment:
What about moving the logic of these two `oauthParams` and `oAuthScope` to some kind of `OAuthUtil` file?

Reply:
done